### ISSUE TITLE
extracting Bottleneck features using pretrained Inceptionv3 - differences between Keras' implementation and Native Tensorflow implementation

### ISSUE THREAD

**Issue by prashanthdumpuri** [keras-team/keras#8418]
(Apologies for the long post)

All,

I want to use the bottleneck features from a pretrained Inceptionv3 model to predict classification for my input images. Before training a model and predicting classification, I tried 3 different approaches for extracting the bottleneck features.

My 3 approaches yielded different bottleneck features (not just in values but even the size was different). 

1. Size of my bottleneck features from Approach 1 and 2: <number of input images x 3 x 3 x 2048>
    Size of my bottleneck features from Approach 3: <number of input images x 2048>

    Why are the sizes different between the Keras based Inceptionv3 model and the native Tensorflow model? My guess is that when I say include_top=False in Keras, I'm not extracting the 'pool_3/_reshape:0' layer. Is this correct? If yes, how do I extract the 'pool_3/_reshape:0' layer in Keras? If my guess is incorrect, what 'am I missing?

2. I compared the bottleneck feature values from Approach 1 and 2 and they were significantly different. I think I'm feeding it the same input images because I resize and rescale my images before I even read it as input for my script. I have no options for my ImageDataGenerator in Approach 1 and according to the documentation for that function all the default values do not change my input image. I have set shuffle to false so I assumed that predict_generator and predict are reading images in the same order. What 'am I missing?
--------------------------------------
Please note:

My inputs images are in RGB format (so number of channels = 3) and I resized all of them to 150x150. I used the preprocess_input function in [inceptionv3.py](https://github.com/fchollet/keras/blob/master/keras/applications/inception_v3.py) to preprocess all my images. 

``` 
    def preprocess_input(image):
    image /= 255.
    image -= 0.5
    image *= 2.
    return image
```
--------------------------------------
## Approach 1: Used Keras with tensorflow as backend, an ImageDataGenerator to read my data and model.predict_generator to compute bottleneck features
I followed the [example](https://blog.keras.io/building-powerful-image-classification-models-using-very-little-data.html) (Section _Using the bottleneck features of a pre-trained network: 90% accuracy in a minute_) from Keras' blog. Instead of VGG model listed there I used Inceptionv3. Below is the snippet of code I used

(code not shown here but what i did before the code below) : read all input images, resize to 150x150x3, rescale according to the preprocessing_input function mentioned above, save the resized and rescaled images

``` 
    train_datagen = ImageDataGenerator() 
    train_generator = train_datagen.flow_from_directory(my_input_dir, target_size=(150,150),shuffle=False, batch_size=16)

    # get bottleneck features
    # use pre-trained model and exclude top layer - which is used for classification
    pretrained_model = InceptionV3(include_top=False, weights='imagenet', input_shape=(150,150,3))
    bottleneck_features_train_v1 = pretrained_model.predict_generator(train_generator,len(train_generator.filenames)//16)
```

## Approach 2: Used Keras with tensorflow as backend, my own reader and model.predict to compute bottleneck features
Only difference between this approach and earlier one is that I used my own reader to read the input images.
(code not shown here but what i did before the code below) : read all input images, resize to 150x150x3, rescale according to the preprocessing_input function mentioned above, save the resized and rescaled images

``` 
    # inputImages is a numpy array of size <number of input images x 150 x 150 x 3>
    inputImages = readAllJPEGsInFolderAndMergeAsRGB(my_input_dir)

    # get bottleneck features
    # use pre-trained model and exclude top layer - which is used for classification
    pretrained_model = InceptionV3(include_top=False, weights='imagenet', input_shape=(img_width, img_height, 3))
    bottleneck_features_train_v2 = pretrained_model.predict(trainData.images,batch_size=16)
```
## Approach 3: Used tensorflow (NO KERAS) compute bottleneck features

I followed [retrain.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/image_retraining/retrain.py) to extract bottleneck features for my input images. Please note that that the weights from that script can be obtained from (http://download.tensorflow.org/models/image/imagenet/inception-2015-12-05.tgz)

As mentioned in that example, I used the bottleneck_tensor_name = 'pool_3/_reshape:0' as the layer to extract and compute bottleneck features. Similar to the first 2 approaches, I used resized and rescaled images as input to the script and I called this feature list bottleneck_features_train_v3

Thank you so much

-------------------------------------------------

P.S. I have posted this question in [stackoverflow](https://stackoverflow.com/questions/47166191/extracting-bottleneck-features-using-pretrained-inceptionv3-differences-betwee) too. I will post an answer here if I it is answered on SO.

**Comment by brge17** [comment]
I would recommend doing something like the following:

`
        bottleneck_model = Model(inputs=source_model.input,
            outputs=self.source_model.get_layer(index=[insert number]).output)
`

and then simply call predict_on_batch or predict_generator.

**Comment by prashanthdumpuri** [comment]
Thanks @brge17 . I'm sorry but I have just started using Keras. 

Per your suggestion,  Model will accept my input images as a tensor. Since 'am using Inception V3, my tensor has to be a 4D array of shape number of input images x  image height x image width x 3. 

> outputs=self.source_model.get_layer(index=[insert number]).output)

Do I use the layer from Inception V3 that I want to extract and compute bottleneck features?

Can you please elaborate?

I have a follow up question or two based on your response. Thanks again for your response.

**Comment by brge17** [comment]
Working code is below. Step 1 is to load the Inception V3 model, step 2 is to print it and find where you want the bottleneck features from, step 3 is to perform surgery and make a new network with the same input as Inception V3 and the desired output, and finally step 4 is to predict.

Change the get layer string as needed.

```
from keras.models import Model
from keras.applications import InceptionV3

import numpy as np

# Randomly generated image
dummy_img = np.random.rand(1, 299, 299, 3)

# Source model
model = InceptionV3()
print model.summary()

# Surgery
bottleneck_model = Model(inputs=model.input,
						 outputs=model.get_layer('avg_pool').output)
print bottleneck_model.summary()

# Predict
bottleneck_features = bottleneck_model.predict(dummy_img)
print bottleneck_features.shape
print bottleneck_features
```

**Comment by prashanthdumpuri** [comment]
Thanks @brge17 . Now all 3 of my approaches (Keras' predict and predict_generator and native tensorflow implementation) give me the same bottleneck features.

Looking at the InceptionV3_model.summary(), include_top=False is not the same as InceptionV3_model.get_layer('avg_pool') are not the same. What I need to compute bottleneck features is the output of avg_pool layer.

Thanks again

p.s. can i close the issue? also, for future references do i post similar questions on this site or on stackoverflow?

**Comment by k0nd0r** [comment]
I am trying to do the same thing but I am still getting different bottleneck features from keras (2.1.1) for InceptionV3 and from native tensorflow. The used image as well as the graph definition file can be found at http://download.tensorflow.org/models/image/imagenet/inception-2015-12-05.tgz . 
How did you manage to get the same results? Here is my python code: 
```
import tensorflow as tf
import numpy as np
from keras.applications.inception_v3 import InceptionV3, preprocess_input
from keras.preprocessing import image
from keras.models import Model

def create_graph():
	with tf.gfile.FastGFile('inception/classify_image_graph_def.pb', 'rb') as f:
		graph_def = tf.GraphDef()
		graph_def.ParseFromString(f.read())
	_ = tf.import_graph_def(graph_def, name='')

def extract_features_tf(image_path):
	create_graph()
	with tf.Session() as sess:
		next_to_last_tensor = sess.graph.get_tensor_by_name('pool_3/_reshape:0')
		image_data = tf.gfile.FastGFile(image_path, 'rb').read()
		predictions = sess.run(next_to_last_tensor, {'DecodeJpeg/contents:0': image_data})
		return np.squeeze(predictions)

def extract_features_keras(image_path):
	base_model = InceptionV3()
	model = Model(inputs=base_model.input, outputs=base_model.get_layer('avg_pool').output)	
	img = image.load_img(image_path, target_size=(299, 299))
	x = image.img_to_array(img)
	x = np.expand_dims(x, axis=0)
	x = preprocess_input(x)
	predictions = model.predict(x)
	return np.squeeze(predictions)

print extract_features_tf("panda.jpg")
[ 0.21214356  0.04288584  0.1422075  ...,  0.09034983  0.01486579
  0.13966659]
print extract_features_keras("panda.jpg")
[ 0.2067015   0.10939689  0.08871846 ...,  0.51381278  0.16807012
  0.23888288]
```

Thanking you in advance.

**Comment by prashanthdumpuri** [comment]
@k0nd0r couple things to try
1. Try the 2016 weights for tensorflow 
2. You are preprocessing your input for keras' bottleneck features but are not doing that for the  tensorflow. (preprocessing in keras scales and normalizes your image to conform to imagenet images). 
Hope that helps.

**Comment by DrChrisLevy** [comment]
@prashanthdumpuri , if you could provide the code showing how you got the same embedding vectors from the different approaches that would be appreciated. Have some legacy code for which uses old native TF implementation of inception and want to update to tf.keras.applications.inception etc. Cheers.
